ruby/examples/lib/register_user_with_form.rb

```diff
@@ -0,0 +1,39 @@
+require 'active_model'
+
+class RegisterUserWithForm
+  Request = Struct.new(:email, :name, :password)
+  Response = Struct.new(:form)
+
+  class Response::Form
+    include ActiveModel::Model
+
+    attr_reader *Request.members
+    attr_reader :errors
+
+    def initialize(email:, name:, password:, errors:)
+      @email = email
+      @name = name
+      @password = password
+      @errors = errors
+    end
+  end
+
+  attr_reader :gateway
+
+  def initialize(gateway: Gateway.new)
+    @gateway = gateway
+  end
+
+  def register(request)
+    errors = gateway.create_user(request.to_h)
+    form = Response::Form.new(request.to_h.merge(errors: errors))
+    Response.new(form)
+  end
+
+  class Gateway
+    def create_user(attributes)
+      order = Spree::User.create(attributes)
+      order.errors
```

Comment:
I also thing Validation is a totally different concern to database access.

This is a dependency that points from the Database logic to the Business logic... In order for the database to know what "is valid" it must therefore know about all possible use cases of it's data structures. This is a violation of the open / closed principle which leads to all of the design smells viscosity, rigidity, immobility, fragility.

> And when a system is fragile and rigid, business people and customers can only come to one conclusion: the developers have lost control of their software, and don't know what they are doing.